File: android/src/main/java/org/maplibre/reactnative/components/annotation/MLRNPointAnnotationManager.java
package org.maplibre.reactnative.components.annotation;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import org.maplibre.reactnative.components.AbstractEventEmitter;
import org.maplibre.reactnative.events.constants.EventKeys;
import org.maplibre.reactnative.utils.GeoJSONUtils;

import java.util.Map;

public class MLRNPointAnnotationManager extends AbstractEventEmitter<MLRNPointAnnotation> {
    public static final String REACT_CLASS = "MLRNPointAnnotation";

    public MLRNPointAnnotationManager(ReactApplicationContext reactApplicationContext) {
        super(reactApplicationContext);
    }

    @Override
    public Map<String, String> customEvents() {
        return MapBuilder.<String, String>builder()
                .put(EventKeys.POINT_ANNOTATION_SELECTED, "onMapboxPointAnnotationSelected")
                .put(EventKeys.POINT_ANNOTATION_DESELECTED, "onMapboxPointAnnotationDeselected")
                .put(EventKeys.POINT_ANNOTATION_DRAG_START, "onMapboxPointAnnotationDragStart")
                .put(EventKeys.POINT_ANNOTATION_DRAG, "onMapboxPointAnnotationDrag")
                .put(EventKeys.POINT_ANNOTATION_DRAG_END, "onMapboxPointAnnotationDragEnd")
                .build();
    }

    //region React Methods
    public static final int METHOD_REFRESH = 2;

    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        return MapBuilder.<String, Integer>builder()
                .put("refresh", METHOD_REFRESH)
                .build();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected MLRNPointAnnotation createViewInstance(ThemedReactContext reactContext) {
        return new MLRNPointAnnotation(reactContext, this);
    }

    @ReactProp(name="id")
    public void setId(MLRNPointAnnotation annotation, String id) {
        annotation.setID(id);
    }

    @ReactProp(name="coordinate")
    public void setCoordinate(MLRNPointAnnotation annotation, String geoJSONStr) {
        annotation.setCoordinate(GeoJSONUtils.toPointGeometry(geoJSONStr));
    }

    @ReactProp(name="anchor")
    public void setAnchor(MLRNPointAnnotation annotation, ReadableMap map) {
        annotation.setAnchor((float) map.getDouble("x"), (float) map.getDouble("y"));
    }

    @ReactProp(name="draggable")
    public void setDraggable(MLRNPointAnnotation annotation, Boolean draggable) {
        annotation.setDraggable(draggable);
    }

    @Override
    public void receiveCommand(MLRNPointAnnotation annotation, int commandID, @Nullable ReadableArray args) {
        switch (commandID) {
            case METHOD_REFRESH:
                annotation.refresh();
                break;
        }
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/annotations/MLRNPointAnnotationManager.kt
package org.maplibre.reactnative.components.annotations

import android.view.View
import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.MLRNPointAnnotationManagerDelegate
import com.facebook.react.viewmanagers.MLRNPointAnnotationManagerInterface

@ReactModule(name = MLRNPointAnnotationManager.REACT_CLASS)
class MLRNPointAnnotationManager(private val reactApplicationContext: ReactApplicationContext) :
    ViewGroupManager<MLRNPointAnnotation>(),
    MLRNPointAnnotationManagerInterface<MLRNPointAnnotation> {

    private val delegate: MLRNPointAnnotationManagerDelegate<MLRNPointAnnotation, MLRNPointAnnotationManager> =
        MLRNPointAnnotationManagerDelegate(this)

    override fun getDelegate(): ViewManagerDelegate<MLRNPointAnnotation> = delegate

    companion object {
        const val REACT_CLASS: String = "MLRNPointAnnotation"
    }

    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext): MLRNPointAnnotation {
        return MLRNPointAnnotation(reactContext)
    }

    override fun addView(parent: MLRNPointAnnotation, child: View, index: Int) {
        parent.addView(child, index)
    }

    override fun removeViewAt(parent: MLRNPointAnnotation, index: Int) {
        parent.removeViewAt(index)
    }

    override fun getChildCount(parent: MLRNPointAnnotation): Int {
        return parent.childCount
    }

    override fun getChildAt(parent: MLRNPointAnnotation, index: Int): View? {
        return parent.getChildAt(index)
    }

    @ReactProp(name = "id")
    override fun setId(annotation: MLRNPointAnnotation, id: String?) {
        id?.let { annotation.setID(it) }
    }

    @ReactProp(name = "title")
    override fun setTitle(annotation: MLRNPointAnnotation, title: String?) {
        annotation.setTitle(title)
    }

    @ReactProp(name = "snippet")
    override fun setSnippet(annotation: MLRNPointAnnotation, snippet: String?) {
        annotation.setSnippet(snippet)
    }

    @ReactProp(name = "selected")
    override fun setSelected(annotation: MLRNPointAnnotation, selected: Boolean) {
        // Selection is handled by the map view
    }

    @ReactProp(name = "lngLat")
    override fun setLngLat(annotation: MLRNPointAnnotation, lngLat: Dynamic) {
        if (lngLat.type == ReadableType.Array) {
            val arr = lngLat.asArray()
            if (arr != null && arr.size() >= 2) {
                annotation.setLngLat(doubleArrayOf(arr.getDouble(0), arr.getDouble(1)))
            }
        }
    }

    @ReactProp(name = "anchor")
    override fun setAnchor(annotation: MLRNPointAnnotation, map: ReadableMap?) {
        if (map != null) {
            annotation.setAnchor(map.getDouble("x").toFloat(), map.getDouble("y").toFloat())
        }
    }

    @ReactProp(name = "draggable")
    override fun setDraggable(annotation: MLRNPointAnnotation, draggable: Boolean) {
        annotation.setDraggable(draggable)
    }

    override fun refresh(annotation: MLRNPointAnnotation) {
        annotation.refresh()
    }
}
